File: training/recruit_features.py
"""Shared recruit-feature extractor for trajectory + freshman-impact models.

Wire-locked feature names. The freshman-impact prior model (Phase 6) will
consume this same module so a single recruit-feature change benefits both.

Coverage caveat: only ~7% of trajectory training rows currently have recruit
data. Ingested classes: 2022, 2023, 2024, 2025, 2026 (2026 not yet played).
Class-of-2021 and earlier (most upperclassmen returners in 2024-25 and
2025-26 trajectory pairs) have no `recruits` row, so the model sees
`recruit_is_ranked=0` and sentinel values for those rows. LightGBM fits
a separate split on the unranked-majority cohort. Historical recruit
backfill (class-of-2021 and earlier) is the obvious lever to expand
coverage on a separate PR.
"""

from __future__ import annotations

from typing import Optional

import numpy as np
import pandas as pd

# Feature names in canonical order. MUST match the Rust-side
# RECRUIT_FEATURE_NAMES in cstat-core::recruit_features. The training pipeline
# and the inference path share this single source of truth via the
# trajectory_model_meta.json contract; validators hard-fail on drift.
RECRUIT_FEATURE_NAMES = [
    "recruit_is_ranked",        # 0/1 flag — explicit "we have a 247 row" signal
    "recruit_composite_rank",   # 247 overall rank (1=best). -1 if unranked.
    "recruit_composite_rating", # 247 0–1 rating score. 0.0 if missing.
    "recruit_star_rating",      # 2–5. 0 if unranked.
    "recruit_position_rank",    # Rank within position. -1 if missing.
    "recruit_rank_movement",    # previous_rank - composite_rank. 0 if either missing.
    "recruit_height_in",        # Parsed from text (e.g., "6-7" → 79). 0 if missing.
    "recruit_weight_lb",        # 0 if missing.
    "recruit_bmi_proxy",        # 703 * weight / height^2 (BMI formula). 0 if either missing.
    "recruit_position_code",    # PG=0, SG=1, SF=2, PF=3, C=4, CG=5, other=-1.
    "years_since_recruit",      # s_n - recruit.year. -1 if no recruit row.
]
N_RECRUIT_FEATURES = len(RECRUIT_FEATURE_NAMES)

# Position taxonomy from 247 (verified against ingested 2022–2026 classes).
# CG = Combo Guard (PG/SG hybrid). Anything outside this set maps to -1.
POSITION_CODES = {
    "PG": 0,
    "SG": 1,
    "SF": 2,
    "PF": 3,
    "C":  4,
    "CG": 5,
}


def parse_height_text(s: Optional[str]) -> Optional[int]:
    """247's height field is `feet-inches` text (e.g., `6-7`). Returns
    total inches or None if unparseable. Empirically all 1,147 resolved
    recruits parse cleanly (range 5-6 through 7-3)."""
    if s is None or pd.isna(s):
        return None
    try:
        feet_str, inches_str = s.split("-", 1)
        feet, inches = int(feet_str), int(inches_str)
        if feet < 4 or feet > 8 or inches < 0 or inches > 11:
            return None
        return feet * 12 + inches
    except (ValueError, AttributeError):
        return None


def encode_position(p: Optional[str]) -> int:
    """Map 247 position string to an integer code; -1 for missing/unknown.

    LightGBM treats this as a continuous feature, which is fine — the
    cardinality is small (6 values) and the model can split each off
    individually. If we ever expand to more positions or want true
    categorical handling, switch to LightGBM's `categorical_feature`.
    """
    if p is None or pd.isna(p):
        return -1
    return POSITION_CODES.get(p.strip().upper(), -1)


def derive_recruit_features(df: pd.DataFrame, prior_season_col: str = "s_n") -> pd.DataFrame:
    """Derive the locked RECRUIT_FEATURE_NAMES columns from raw recruit
    fields. Mutates the DataFrame in place and returns it for chaining.

    Input columns (callers SELECT these explicitly — the column shapes
    diverge between the trajectory query, which aliases `rec.year` as
    `recruit_year_raw`, and the freshman query, which aliases `r.year`
    three different ways — so a single SQL helper wouldn't fit both):
        recruit_composite_rank_raw, recruit_composite_rating_raw,
        recruit_star_rating_raw, recruit_position_rank_raw,
        recruit_previous_rank_raw, recruit_height_raw, recruit_weight_raw,
        recruit_position_raw, recruit_year_raw

    Plus `prior_season_col` (default `s_n`) for the `years_since_recruit`
    derivation.

    NULL handling:
      - `recruit_is_ranked = 1 if recruit_year_raw is not null else 0`
        (the year column is NOT NULL in `recruits`, so this is the cleanest
        existence check across the LEFT JOIN).
      - All rank-like fields use -1 sentinel for missing.
      - All rate/count fields use 0.0 for missing.
      - Position uses -1 for missing/unknown.
    """
    has_recruit = df["recruit_year_raw"].notna()
    df["recruit_is_ranked"] = has_recruit.astype(int)

    # Rank fields → -1 sentinel.
    df["recruit_composite_rank"] = (
        df["recruit_composite_rank_raw"].fillna(-1).astype(int)
    )
    df["recruit_position_rank"] = (
        df["recruit_position_rank_raw"].fillna(-1).astype(int)
    )

    # Rating / star → 0.0 / 0.
    df["recruit_composite_rating"] = (
        df["recruit_composite_rating_raw"].fillna(0.0).astype(float)
    )
    df["recruit_star_rating"] = (
        df["recruit_star_rating_raw"].fillna(0).astype(int)
    )

    # Rank movement: positive = climbed (previous rank was higher number),
    # negative = fell. Set to 0 when either rank is missing — neutral signal.
    prev = df["recruit_previous_rank_raw"]
    curr = df["recruit_composite_rank_raw"]
    movement = prev - curr
    df["recruit_rank_movement"] = movement.fillna(0).astype(int)

    # Height parse + weight.
    df["recruit_height_in"] = (
        df["recruit_height_raw"].apply(parse_height_text).fillna(0).astype(int)
    )
    df["recruit_weight_lb"] = (
        df["recruit_weight_raw"].fillna(0).astype(int)
    )

    # BMI proxy using the standard imperial formula. Use safe division so the
    # 0-height fallback doesn't NaN out. Vectorised because `np.where` is
    # cheap and we want to avoid surprise division-by-zero warnings.
    h = df["recruit_height_in"].astype(float)
    w = df["recruit_weight_lb"].astype(float)
    safe_h2 = np.where(h > 0, h * h, 1.0)  # placeholder denom; mask kills it
    bmi = 703.0 * w / safe_h2
    df["recruit_bmi_proxy"] = np.where((h > 0) & (w > 0), bmi, 0.0)

    # Position code.
    df["recruit_position_code"] = (
        df["recruit_position_raw"].apply(encode_position).astype(int)
    )

    # Years since recruit. Sentinel -1 when no recruit row (so the model
    # can split on "did they have a recruit row at all" via is_ranked AND
    # the years_since signal at the same time).
    years_since = df[prior_season_col] - df["recruit_year_raw"]
    df["years_since_recruit"] = years_since.fillna(-1).astype(int)

    return df

File: crates/cstat-core/src/recruit_features.rs
//! Shared recruit-feature block consumed by the Phase 5c trajectory model
//! and (next) the freshman-impact prior model. Mirrors
//! `training/recruit_features.py` — feature order, sentinel conventions,
//! and the position-code taxonomy must match exactly. The
//! `trajectory_model_meta.json` validator at boot enforces the contract.
//!
//! Coverage caveat: class-of-2022 through 2026 are ingested today, but
//! the majority of trajectory-model rows are upperclassmen from earlier
//! classes that have no recruit row and fall into the
//! `recruit_is_ranked=0` bucket via sentinel encoding. LightGBM fits a
//! separate branch on the unranked-majority cohort. Historical backfill
//! (class-of-2021 and earlier) is a follow-up PR.
//!
//! Feature shape (11 cols, locked order — must equal Python
//! `RECRUIT_FEATURE_NAMES` in `training/recruit_features.py`):
//!   - `recruit_is_ranked` (0/1 — existence flag, drives a model branch)
//!   - `recruit_composite_rank` (i32; -1 sentinel)
//!   - `recruit_composite_rating` (f32; 0.0 missing)
//!   - `recruit_star_rating` (i32; 0 missing)
//!   - `recruit_position_rank` (i32; -1 sentinel)
//!   - `recruit_rank_movement` (i32; previous − current, 0 if either missing)
//!   - `recruit_height_in` (i32; parsed from "feet-inches", 0 if missing)
//!   - `recruit_weight_lb` (i32; 0 if missing)
//!   - `recruit_bmi_proxy` (f32; standard imperial BMI, 0.0 if either missing)
//!   - `recruit_position_code` (i32; PG=0..C=4, CG=5, other=-1)
//!   - `years_since_recruit` (i32; prior_season − recruit.year, -1 if no row)

/// Number of features the recruit block contributes. Wire-locked to
/// `RECRUIT_FEATURE_NAMES` and the Python-side
/// `training/recruit_features.py::N_RECRUIT_FEATURES`.
pub const RECRUIT_NUM_FEATURES: usize = 11;

/// Feature names in the exact order producers/consumers must use.
pub const RECRUIT_FEATURE_NAMES: [&str; RECRUIT_NUM_FEATURES] = [
    "recruit_is_ranked",
    "recruit_composite_rank",
    "recruit_composite_rating",
    "recruit_star_rating",
    "recruit_position_rank",
    "recruit_rank_movement",
    "recruit_height_in",
    "recruit_weight_lb",
    "recruit_bmi_proxy",
    "recruit_position_code",
    "years_since_recruit",
];

/// Raw recruit row, captured from a LEFT JOIN against `recruits`. All fields
/// are `Option<…>` because the join is LEFT (most pre-2024 returners have
/// no recruit row); when `year` is `None` the entire block is treated as
/// "no recruit data" (`recruit_is_ranked=0` and sentinel values throughout).
#[derive(Debug, Clone, Default)]
pub struct RecruitFeatureRow {
    pub composite_rank: Option<i32>,
    pub composite_rating: Option<f64>,
    pub star_rating: Option<i32>,
    pub position_rank: Option<i32>,
    pub previous_rank: Option<i32>,
    pub height: Option<String>,
    pub weight: Option<i32>,
    pub position: Option<String>,
    /// Recruiting class year. `Some(y)` is the existence signal — when this
    /// is `None`, we treat all other fields as missing too (the LEFT JOIN
    /// returned no row).
    pub year: Option<i32>,
}

/// Parse 247's `feet-inches` height string into total inches. Returns
/// `None` on unparseable / out-of-range input. Empirically all ingested
/// recruits (n=1,147 with a cstat join) parse cleanly in range `5-6`
/// through `7-3`.
pub fn parse_height_inches(s: Option<&str>) -> Option<i32> {
    let s = s?;
    let (feet_str, inches_str) = s.split_once('-')?;
    let feet: i32 = feet_str.trim().parse().ok()?;
    let inches: i32 = inches_str.trim().parse().ok()?;
    if !(4..=8).contains(&feet) || !(0..=11).contains(&inches) {
        return None;
    }
    Some(feet * 12 + inches)
}

/// Map a 247 position string to its integer code. `None` / unknown → -1.
/// MUST match the Python `POSITION_CODES` dict in `recruit_features.py`.
pub fn position_code(p: Option<&str>) -> i32 {
    match p.map(|x| x.trim().to_ascii_uppercase()) {
        Some(ref s) if s == "PG" => 0,
        Some(ref s) if s == "SG" => 1,
        Some(ref s) if s == "SF" => 2,
        Some(ref s) if s == "PF" => 3,
        Some(ref s) if s == "C" => 4,
        Some(ref s) if s == "CG" => 5,
        _ => -1,
    }
}

/// Build the 11-element recruit feature block. `prior_season` is the
/// trajectory model's `s_n` (i.e., the season we're predicting *from*) —
/// `years_since_recruit = prior_season - recruit.year`. When the recruit
/// row is absent (`row.year.is_none()`), every slot uses its sentinel.
pub fn build_recruit_feature_block(
    row: &RecruitFeatureRow,
    prior_season: i32,
) -> [f32; RECRUIT_NUM_FEATURES] {
    let is_ranked = row.year.is_some();

    let composite_rank = row.composite_rank.unwrap_or(-1);
    let composite_rating = row.composite_rating.unwrap_or(0.0);
    let star_rating = row.star_rating.unwrap_or(0);
    let position_rank = row.position_rank.unwrap_or(-1);

    // Rank movement: positive = climbed (previous rank was a higher number),
    // negative = fell. Both ranks must be present; 0 is the neutral sentinel.
    let rank_movement = match (row.previous_rank, row.composite_rank) {
        (Some(prev), Some(curr)) => prev - curr,
        _ => 0,
    };

    let height_in = parse_height_inches(row.height.as_deref()).unwrap_or(0);
    let weight_lb = row.weight.unwrap_or(0);
    let bmi_proxy = if height_in > 0 && weight_lb > 0 {
        703.0 * (weight_lb as f64) / ((height_in as f64) * (height_in as f64))
    } else {
        0.0
    };

    let pos_code = position_code(row.position.as_deref());
    let years_since = match row.year {
        Some(y) => prior_season - y,
        None => -1,
    };

    // Order MUST match RECRUIT_FEATURE_NAMES.
    [
        if is_ranked { 1.0 } else { 0.0 },
        composite_rank as f32,
        composite_rating as f32,
        star_rating as f32,
        position_rank as f32,
        rank_movement as f32,
        height_in as f32,
        weight_lb as f32,
        bmi_proxy as f32,
        pos_code as f32,
        years_since as f32,
    ]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn height_parser() {
        assert_eq!(parse_height_inches(Some("6-7")), Some(79));
        assert_eq!(parse_height_inches(Some("5-10")), Some(70));
        assert_eq!(parse_height_inches(Some("7-3")), Some(87));
        assert_eq!(parse_height_inches(Some("6-12")), None); // inches out of range
        assert_eq!(parse_height_inches(Some("9-0")), None); // feet out of range
        assert_eq!(parse_height_inches(Some("garbage")), None);
        assert_eq!(parse_height_inches(None), None);
    }

    #[test]
    fn position_codes_match_python() {
        // These integers are the contract with `training/recruit_features.py`.
        assert_eq!(position_code(Some("PG")), 0);
        assert_eq!(position_code(Some("SG")), 1);
        assert_eq!(position_code(Some("SF")), 2);
        assert_eq!(position_code(Some("PF")), 3);
        assert_eq!(position_code(Some("C")), 4);
        assert_eq!(position_code(Some("CG")), 5);
        assert_eq!(position_code(Some("pg")), 0); // case-insensitive
        assert_eq!(position_code(Some("  SF  ")), 2); // trims whitespace
        assert_eq!(position_code(Some("G")), -1); // unknown
        assert_eq!(position_code(None), -1);
    }

    #[test]
    fn block_layout_full_row() {
        // 247-style recruit: ranked #5, 6'7", 215 lb, SF, class-of-2025,
        // moved up 3 spots from previous list (prev rank 8). Projecting
        // from cstat-season 2026 = 1 year past freshman.
        let row = RecruitFeatureRow {
            composite_rank: Some(5),
            composite_rating: Some(0.9852),
            star_rating: Some(5),
            position_rank: Some(2),
            previous_rank: Some(8),
            height: Some("6-7".into()),
            weight: Some(215),
            position: Some("SF".into()),
            year: Some(2025),
        };
        let v = build_recruit_feature_block(&row, 2026);

        let idx = |name: &str| {
            RECRUIT_FEATURE_NAMES
                .iter()
                .position(|n| *n == name)
                .unwrap()
        };
        assert_eq!(v[idx("recruit_is_ranked")], 1.0);
        assert_eq!(v[idx("recruit_composite_rank")], 5.0);
        assert!((v[idx("recruit_composite_rating")] - 0.9852).abs() < 1e-4);
        assert_eq!(v[idx("recruit_star_rating")], 5.0);
        assert_eq!(v[idx("recruit_position_rank")], 2.0);
        assert_eq!(v[idx("recruit_rank_movement")], 3.0); // 8 - 5
        assert_eq!(v[idx("recruit_height_in")], 79.0); // 6'7"
        assert_eq!(v[idx("recruit_weight_lb")], 215.0);
        // BMI = 703 * 215 / 79^2 ≈ 24.21
        assert!((v[idx("recruit_bmi_proxy")] - 24.21).abs() < 0.05);
        assert_eq!(v[idx("recruit_position_code")], 2.0); // SF
        assert_eq!(v[idx("years_since_recruit")], 1.0); // 2026 - 2025
    }

    #[test]
    fn missing_row_becomes_sentinels() {
        let row = RecruitFeatureRow::default(); // year is None → no recruit
        let v = build_recruit_feature_block(&row, 2026);

        let idx = |name: &str| {
            RECRUIT_FEATURE_NAMES
                .iter()
                .position(|n| *n == name)
                .unwrap()
        };
        assert_eq!(v[idx("recruit_is_ranked")], 0.0);
        assert_eq!(v[idx("recruit_composite_rank")], -1.0);
        assert_eq!(v[idx("recruit_composite_rating")], 0.0);
        assert_eq!(v[idx("recruit_star_rating")], 0.0);
        assert_eq!(v[idx("recruit_position_rank")], -1.0);
        assert_eq!(v[idx("recruit_rank_movement")], 0.0);
        assert_eq!(v[idx("recruit_height_in")], 0.0);
        assert_eq!(v[idx("recruit_weight_lb")], 0.0);
        assert_eq!(v[idx("recruit_bmi_proxy")], 0.0);
        assert_eq!(v[idx("recruit_position_code")], -1.0);
        assert_eq!(v[idx("years_since_recruit")], -1.0);
    }

    #[test]
    fn previous_rank_missing_movement_zero() {
        // Ranked freshman without previous-rank history (44pct of ranked
        // recruits in cstat data, per recruits-table coverage check).
        let row = RecruitFeatureRow {
            composite_rank: Some(50),
            composite_rating: Some(0.9),
            star_rating: Some(4),
            position_rank: Some(8),
            previous_rank: None,
            height: Some("6-3".into()),
            weight: Some(195),
            position: Some("PG".into()),
            year: Some(2024),
        };
        let v = build_recruit_feature_block(&row, 2025);
        let movement_idx = RECRUIT_FEATURE_NAMES
            .iter()
            .position(|n| *n == "recruit_rank_movement")
            .unwrap();
        assert_eq!(v[movement_idx], 0.0);
    }
}
